src/app/features/search-bar/search-bar.component.ts

```diff
@@ -0,0 +1,250 @@
+import {
+  Component,
+  ElementRef,
+  EventEmitter,
+  Input,
+  OnDestroy,
+  Output,
+  ViewChild,
+  AfterViewInit,
+  ChangeDetectionStrategy,
+} from '@angular/core';
+import { T } from '../../t.const';
+import { FormControl } from '@angular/forms';
+import { BehaviorSubject, Observable } from 'rxjs';
+import { tap, map, take, debounceTime, startWith } from 'rxjs/operators';
+import { TaskService } from '../tasks/task.service';
+import { Router } from '@angular/router';
+import { TODAY_TAG } from '../tag/tag.const';
+import { Project } from '../project/project.model';
+import { Tag } from '../tag/tag.model';
+import { ProjectService } from '../project/project.service';
+import { TagService } from '../tag/tag.service';
+import { Task } from '../tasks/task.model';
+import { blendInOutAnimation } from 'src/app/ui/animations/blend-in-out.ani';
+import { AnimationEvent } from '@angular/animations';
+import { SearchItem, SearchQueryParams } from './search-bar.model';
+import { getWorklogStr } from '../../util/get-work-log-str';
+
+@Component({
+  selector: 'search-bar',
+  templateUrl: './search-bar.component.html',
+  styleUrls: ['./search-bar.component.scss'],
+  animations: [blendInOutAnimation],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class SearchBarComponent implements AfterViewInit, OnDestroy {
+  @Input() isElevated: boolean = false;
+  @Output() blurred: EventEmitter<any> = new EventEmitter();
+
+  @ViewChild('inputEl') inputEl!: ElementRef;
+  @ViewChild('searchForm') searchForm!: ElementRef;
+
+  T: typeof T = T;
+  isLoading$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);
+  taskSuggestionsCtrl: FormControl = new FormControl();
+  filteredIssueSuggestions$: Observable<SearchItem[]> = new Observable();
+  isArchivedTasks: boolean = false;
+  tooManyResults: boolean = false;
+
+  private _attachKeyDownHandlerTimeout?: number;
+  private _tasks: Task[] = [];
+  private _projects: Project[] = [];
+  private _tags: Tag[] = [];
+  private _searchableItems: SearchItem[] = [];
+
+  constructor(
+    private _taskService: TaskService,
+    private _projectService: ProjectService,
+    private _tagService: TagService,
+    private _router: Router,
+  ) {
+    this._taskService.allTasks$.pipe(take(1)).subscribe((t) => (this._tasks = t));
+    this._projectService.list$.pipe(take(1)).subscribe((p) => (this._projects = p));
+    this._tagService.tags$.pipe(take(1)).subscribe((t) => (this._tags = t));
+
+    this._mapTasksToSearchItems();
+  }
+
+  private _mapTasksToSearchItems(): void {
```

Comment:
Just as a note: When you're leaning to the functional side of programming (like I do ;)) then it's considered cleaner to use params and return values rather than having a function that reads and assigns class properties.

Reply:
Interesting topic. I like functional programming, too, but for me it isn't always easy to find the line between when to "use it" and when not. Especially in an OO framework like Angular. In this specific case I could pass the _tasks object but later on in _getArchivedDate() I'll be needing those tasks to be a class property anyways. At least I don't see a clean way of passing them unless to add the optionSelected listener programmatically. Then _searchableItems could be returned and passed to _setUpFilter() but the latter will only be called later on in ngAfterViewInit. For this particular case I don't think it'd be cleaner... Maybe I just don't get it, let me know what you think so I can improve :) and please note, I'm not arguing, just find the topic interesting and like to talk about it :)

Reply:
I am not against OO or using class properties. I was more thinking about doing it like this:
```TypeScript
    this._searchableItems = this._mapTasksToSearchItems(this._tasks);
```

Personally I think it's better because you don't need to go and read the more specific mapTasksToSearchItems, to see what variables are needed/affected. Gives you the info what value is set initially, which I think is good to know when you're trying to get an overview of what is going on here.

Reply:
oh ok, simply like that. Sure I'll fix that. Got a bit carried away :)